File: internal/tunnel/protocol.go
package tunnel

import "encoding/json"

// GatewayMessage is a JSON message exchanged over the encrypted channel.
type GatewayMessage struct {
	SessionID string          `json:"session_id,omitempty"`
	EventID   string          `json:"event_id,omitempty"`
	StreamID  string          `json:"stream_id,omitempty"`
	Type      string          `json:"type"`
	Data      json.RawMessage `json:"data,omitempty"`
}

// Protocol defines the message types exchanged between the ggcode backend
// and mobile clients over the WebSocket tunnel.
//
// Server → Client: event messages (lowercase type)
// Client → Server: command messages (lowercase type)

// Server → Client event types.
const (
	EventConnected          = "connected"
	EventActiveSession      = "active_session"
	EventSessionInfo        = "session_info"
	EventActivity           = "activity"     // main-agent activity text change
	EventUserMessage        = "user_message" // user text from desktop
	EventSystemMessage      = "system_message"
	EventText               = "text"      // streaming text chunk
	EventTextDone           = "text_done" // text stream complete
	EventStatus             = "status"    // agent status change
	EventToolCall           = "tool_call"
	EventToolResult         = "tool_result"
	EventApprovalRequest    = "approval_request"
	EventApprovalResult     = "approval_result"
	EventAskUserRequest     = "ask_user_request"
	EventAskUserResponse    = "ask_user_response"
	EventSubagentSpawn      = "subagent_spawn"
	EventSubagentText       = "subagent_text"
	EventSubagentStatus     = "subagent_status"
	EventSubagentToolCall   = "subagent_tool_call"
	EventSubagentToolResult = "subagent_tool_result"
	EventSubagentComplete   = "subagent_complete"
	EventResumeMiss         = "resume_miss"
	EventSnapshotReset      = "snapshot_reset"
	EventLanguageChange     = "language_change" // bidirectional language sync
	EventError              = "error"
	EventPing               = "ping"
	EventDisconnected       = "disconnected"
)

// Client → Server command types.
const (
	CmdResumeHello      = "resume_hello"
	CmdResumeFrom       = "resume_from"
	CmdMessage          = "message"
	CmdApprovalResponse = "approval_response"
	CmdInterrupt        = "interrupt"
	CmdModeChange       = "mode_change"
	CmdAskUserResponse  = "ask_user_response"
	CmdPong             = "pong"
	CmdLanguageChange   = "language_change"
)

// Agent status values.
const (
	StatusIdle     = "idle"
	StatusBusy     = "busy"
	StatusThinking = "thinking"
	StatusRunning  = "running"
	StatusWaiting  = "waiting" // waiting for approval
	StatusError    = "error"
)

// Mode values.
const (
	ModeSupervised = "supervised"
	ModeAuto       = "auto"
	ModeBypass     = "bypass"
	ModeAutopilot  = "autopilot"
)

// Approval decisions.
const (
	DecisionAllow       = "allow"
	DecisionDeny        = "deny"
	DecisionAlwaysAllow = "always_allow"
)

// Resume mode values used by resume_ack.
const (
	ResumeModeIncremental      = "incremental"
	ResumeModeFullHistory      = "full_history"
	ResumeModeSnapshotRequired = "snapshot_required"
)

// SessionInfoData carries session metadata sent after connection.
type SessionInfoData struct {
	Workspace string `json:"workspace"`
	Model     string `json:"model"`
	Provider  string `json:"provider"`
	Mode      string `json:"mode"`
	Version   string `json:"version"`
	Language  string `json:"language,omitempty"`
	Theme     string `json:"theme,omitempty"`
}

// ActiveSessionData declares the authoritative TUI/GUI session bound to the
// current tunnel room.
type ActiveSessionData struct {
	SessionID string `json:"session_id"`
}

// LanguageChangeData carries a language preference change (bidirectional sync).
type LanguageChangeData struct {
	Language string `json:"language"` // "en" or "zh-CN"
}

// ThemeChangeData carries a theme change (bidirectional sync).
type ThemeChangeData struct {
	Theme string `json:"theme"` // "midnight", "oled", "nord", "rose", "forest", "light"
}

// EventThemeChange is the event name for theme changes across clients.
const EventThemeChange = "theme_change"

// CmdThemeChange is the command name for theme change requests.
const CmdThemeChange = "theme_change"

// ResumeHelloData requests either incremental replay or full session replay.
type ResumeHelloData struct {
	ClientID    string `json:"client_id"`
	SessionID   string `json:"session_id,omitempty"`
	LastEventID string `json:"last_event_id,omitempty"`
}

// ResumeFromData requests replay starting after the caller's last good cursor.
type ResumeFromData struct {
	ClientID    string `json:"client_id"`
	SessionID   string `json:"session_id,omitempty"`
	LastEventID string `json:"last_event_id,omitempty"`
}

// ResumeAckData tells the client how the gateway will recover state.
type ResumeAckData struct {
	ClientID   string `json:"client_id"`
	SessionID  string `json:"session_id,omitempty"`
	ResumeMode string `json:"resume_mode"`
}

// TextData carries a streaming text chunk.
type TextData struct {
	ID    string `json:"id"`             // message ID to group chunks
	Chunk string `json:"chunk"`          // text content
	Done  bool   `json:"done"`           // true on last chunk
	Kind  string `json:"kind,omitempty"` // optional semantic hint for rendering
}

const (
	MessageKindCron         = "cron"
	MessageKindShellCommand = "shell_command"
	MessageKindShellOutput  = "shell_output"
)

// StatusData carries an agent status change.
type StatusData struct {
	Status  string `json:"status"`
	Message string `json:"message,omitempty"`
}

// ActivityData carries the current main-agent activity text.
type ActivityData struct {
	Activity string `json:"activity,omitempty"`
}

// ToolCallData carries a tool call notification.
type ToolCallData struct {
	ToolID      string `json:"tool_id"`
	ToolName    string `json:"tool_name"`
	DisplayName string `json:"display_name,omitempty"`
	Args        string `json:"args,omitempty"`
	Detail      string `json:"detail,omitempty"`
}

// ToolResultData carries a tool result.
type ToolResultData struct {
	ToolID      string `json:"tool_id"`
	ToolName    string `json:"tool_name"`
	Result      string `json:"result,omitempty"`
	Summary     string `json:"summary,omitempty"`
	Payload     string `json:"payload,omitempty"`
	PayloadMode string `json:"payload_mode,omitempty"`
	IsError     bool   `json:"is_error,omitempty"`
}

// ApprovalRequestData carries an approval request to the mobile client.
type ApprovalRequestData struct {
	ID       string `json:"id"`
	ToolName string `json:"tool_name"`
	Input    string `json:"input"`
}

// ApprovalResponseData carries the user's approval decision back.
type ApprovalResponseData struct {
	ID       string `json:"id"`
	Decision string `json:"decision"` // allow/deny/always_allow
}

// MessageData carries a user message.
type MessageData struct {
	Text        string `json:"text"`
	DisplayText string `json:"display_text,omitempty"`
	Kind        string `json:"kind,omitempty"`
}

// ModeChangeData carries a mode change request.
type ModeChangeData struct {
	Mode string `json:"mode"`
}

// ErrorData carries an error.
type ErrorData struct {
	Message string `json:"message"`
	Code    string `json:"code,omitempty"`
}

// ─── Ask User (structured questionnaire) ───

// AskUserRequestData carries a structured questionnaire from the agent.
// This is NOT the same as approval_request — it's a multi-question form
// with single/multi/text question types.
type AskUserRequestData struct {
	ID        string            `json:"id"`
	Title     string            `json:"title,omitempty"`
	Questions []AskUserQuestion `json:"questions"`
}

// AskUserQuestion represents a single question in the questionnaire.
type AskUserQuestion struct {
	ID            string          `json:"id"`
	Prompt        string          `json:"prompt"`
	Kind          string          `json:"kind"` // single/multi/text
	Choices       []AskUserChoice `json:"choices,omitempty"`
	AllowFreeform bool            `json:"allow_freeform,omitempty"`
	Placeholder   string          `json:"placeholder,omitempty"`
}

// AskUserChoice represents a selectable choice.
type AskUserChoice struct {
	ID    string `json:"id"`
	Label string `json:"label"`
}

// AskUserResponseData carries the user's answers back.
type AskUserResponseData struct {
	ID      string          `json:"id"`
	Status  string          `json:"status"` // submitted/cancelled
	Answers []AskUserAnswer `json:"answers,omitempty"`
}

// AskUserAnswer carries a single question answer.
type AskUserAnswer struct {
	QuestionID   string   `json:"question_id"`
	ChoiceIDs    []string `json:"choice_ids,omitempty"`
	FreeformText string   `json:"freeform_text,omitempty"`
}

// ─── Sub-agent / Teammate ───

// SubagentSpawnData notifies mobile that a sub-agent has been spawned.
// Mobile should show a live activity card for this agent.
type SubagentSpawnData struct {
	AgentID  string `json:"agent_id"`
	Name     string `json:"name"`                // e.g. "Researcher", "Coder"
	Task     string `json:"task"`                // brief task description
	Color    string `json:"color,omitempty"`     // e.g. "#4CAF50"
	ParentID string `json:"parent_id,omitempty"` // empty for top-level
}

// SubagentTextData carries streaming text from a sub-agent.
type SubagentTextData struct {
	AgentID string `json:"agent_id"`
	ID      string `json:"id"` // message ID for grouping chunks
	Chunk   string `json:"chunk"`
	Done    bool   `json:"done"`
}

// SubagentStatusData carries status updates for a sub-agent.
type SubagentStatusData struct {
	AgentID string `json:"agent_id"`
	Status  string `json:"status"` // running/waiting_approval/completed/failed
	Message string `json:"message,omitempty"`
}

// SubagentToolCallData carries tool call info from a sub-agent.
type SubagentToolCallData struct {
	AgentID     string `json:"agent_id"`
	ToolID      string `json:"tool_id"`
	ToolName    string `json:"tool_name"`
	DisplayName string `json:"display_name,omitempty"`
	Args        string `json:"args,omitempty"`
	Detail      string `json:"detail,omitempty"`
}

// SubagentToolResultData carries tool result from a sub-agent.
type SubagentToolResultData struct {
	AgentID     string `json:"agent_id"`
	ToolID      string `json:"tool_id"`
	ToolName    string `json:"tool_name"`
	Result      string `json:"result"`
	Summary     string `json:"summary,omitempty"`
	Payload     string `json:"payload,omitempty"`
	PayloadMode string `json:"payload_mode,omitempty"`
	IsError     bool   `json:"is_error"`
}

// SubagentCompleteData notifies that a sub-agent has finished.
type SubagentCompleteData struct {
	AgentID string `json:"agent_id"`
	Name    string `json:"name"`
	Summary string `json:"summary"` // one-line summary of what was done
	Success bool   `json:"success"`
}

File: mobile/flutter/lib/core/models/protocol.dart
import 'dart:convert';

/// Base class for all WebSocket messages
class WsMessage {
  final String? sessionId;
  final String? eventId;
  final String? streamId;
  final String? messageId;
  final int? generation;
  final String type;
  final Map<String, dynamic>? data;

  WsMessage({
    this.sessionId,
    this.eventId,
    this.streamId,
    this.messageId,
    this.generation,
    required this.type,
    this.data,
  });

  String toJson() => jsonEncode({
        'session_id': sessionId,
        'event_id': eventId,
        'stream_id': streamId,
        'message_id': messageId,
        'generation': generation,
        'type': type,
        'data': data,
      });

  static WsMessage fromJson(String jsonStr) {
    final map = jsonDecode(jsonStr) as Map<String, dynamic>;
    return WsMessage(
      sessionId: map['session_id'] as String?,
      eventId: map['event_id'] as String?,
      streamId: map['stream_id'] as String?,
      messageId: map['message_id'] as String?,
      generation: (map['generation'] as num?)?.toInt(),
      type: map['type'] as String,
      data: map['data'] as Map<String, dynamic>?,
    );
  }
}

/// ─── Session Info ───

class SessionInfoData {
  final String workspace;
  final String model;
  final String provider;
  final String mode;
  final String version;
  final String language;
  final String theme;

  SessionInfoData({
    required this.workspace,
    required this.model,
    required this.provider,
    required this.mode,
    required this.version,
    this.language = '',
    this.theme = '',
  });

  factory SessionInfoData.fromJson(Map<String, dynamic> d) => SessionInfoData(
        workspace: d['workspace'] as String? ?? '',
        model: d['model'] as String? ?? '',
        provider: d['provider'] as String? ?? '',
        mode: d['mode'] as String? ?? '',
        version: d['version'] as String? ?? '',
        language: d['language'] as String? ?? '',
        theme: d['theme'] as String? ?? '',
      );
}

/// ─── Text Streaming ───

class TextData {
  final String id;
  final String chunk;
  final bool done;
  final String kind;

  TextData(
      {required this.id,
      required this.chunk,
      required this.done,
      this.kind = ''});

  factory TextData.fromJson(Map<String, dynamic> d) => TextData(
        id: d['id'] as String? ?? '',
        chunk: d['chunk'] as String? ?? '',
        done: d['done'] as bool? ?? false,
        kind: d['kind'] as String? ?? '',
      );
}

/// ─── Status ───

class StatusData {
  final String status;
  final String message;

  StatusData({required this.status, required this.message});

  factory StatusData.fromJson(Map<String, dynamic> d) => StatusData(
        status: d['status'] as String? ?? '',
        message: d['message'] as String? ?? '',
      );
}

class ActivityData {
  final String activity;

  ActivityData({required this.activity});

  factory ActivityData.fromJson(Map<String, dynamic> d) => ActivityData(
        activity: d['activity'] as String? ?? '',
      );
}

class MessageData {
  final String text;
  final String displayText;
  final String kind;
  final String messageId;

  MessageData({
    required this.text,
    this.displayText = '',
    this.kind = '',
    this.messageId = '',
  });

  Map<String, dynamic> toJson() => {
        'text': text,
        'display_text': displayText,
        'kind': kind,
        'message_id': messageId,
      };

  factory MessageData.fromJson(Map<String, dynamic> d) => MessageData(
        text: d['text'] as String? ?? '',
        displayText: d['display_text'] as String? ?? '',
        kind: d['kind'] as String? ?? '',
        messageId: d['message_id'] as String? ?? '',
      );
}

/// ─── Tool Call / Result ───

class ToolCallData {
  final String toolId;
  final String toolName;
  final String displayName;
  final String args;
  final String detail;

  ToolCallData(
      {required this.toolId,
      required this.toolName,
      this.displayName = '',
      required this.args,
      required this.detail});

  factory ToolCallData.fromJson(Map<String, dynamic> d) => ToolCallData(
        toolId: d['tool_id'] as String? ?? '',
        toolName: d['tool_name'] as String? ?? '',
        displayName: d['display_name'] as String? ?? '',
        args: d['args'] as String? ?? '',
        detail: d['detail'] as String? ?? '',
      );
}

class ToolResultData {
  final String toolId;
  final String toolName;
  final String result;
  final String summary;
  final String payload;
  final String payloadMode;
  final bool isError;

  ToolResultData(
      {required this.toolId,
      required this.toolName,
      required this.result,
      this.summary = '',
      this.payload = '',
      this.payloadMode = '',
      required this.isError});

  factory ToolResultData.fromJson(Map<String, dynamic> d) => ToolResultData(
        toolId: d['tool_id'] as String? ?? '',
        toolName: d['tool_name'] as String? ?? '',
        result: d['result'] as String? ?? '',
        summary: d['summary'] as String? ?? '',
        payload: d['payload'] as String? ?? '',
        payloadMode: d['payload_mode'] as String? ?? '',
        isError: d['is_error'] as bool? ?? false,
      );
}

/// ─── Approval ───

class ApprovalRequestData {
  final String id;
  final String toolName;
  final String input;

  ApprovalRequestData(
      {required this.id, required this.toolName, required this.input});

  factory ApprovalRequestData.fromJson(Map<String, dynamic> d) =>
      ApprovalRequestData(
        id: d['id'] as String? ?? '',
        toolName: d['tool_name'] as String? ?? '',
        input: d['input'] as String? ?? '',
      );
}

class ApprovalResultData {
  final String id;
  final String decision;

  ApprovalResultData({required this.id, required this.decision});

  factory ApprovalResultData.fromJson(Map<String, dynamic> d) =>
      ApprovalResultData(
        id: d['id'] as String? ?? '',
        decision: d['decision'] as String? ?? '',
      );
}

/// ─── Ask User (Structured Questionnaire) ───

class AskUserQuestion {
  final String id;
  final String prompt;
  final String kind; // single/multi/text
  final List<AskUserChoice> choices;
  final bool allowFreeform;
  final String placeholder;

  AskUserQuestion({
    required this.id,
    required this.prompt,
    required this.kind,
    this.choices = const [],
    this.allowFreeform = false,
    this.placeholder = '',
  });

  factory AskUserQuestion.fromJson(Map<String, dynamic> d) => AskUserQuestion(
        id: d['id'] as String? ?? '',
        prompt: d['prompt'] as String? ?? '',
        kind: d['kind'] as String? ?? 'text',
        choices: (d['choices'] as List<dynamic>?)
                ?.map((c) => AskUserChoice.fromJson(c as Map<String, dynamic>))
                .toList() ??
            [],
        allowFreeform: d['allow_freeform'] as bool? ?? false,
        placeholder: d['placeholder'] as String? ?? '',
      );
}

class AskUserChoice {
  final String id;
  final String label;

  AskUserChoice({required this.id, required this.label});

  factory AskUserChoice.fromJson(Map<String, dynamic> d) => AskUserChoice(
        id: d['id'] as String? ?? '',
        label: d['label'] as String? ?? '',
      );
}

class AskUserRequestData {
  final String id;
  final String title;
  final List<AskUserQuestion> questions;

  AskUserRequestData({
    required this.id,
    required this.title,
    required this.questions,
  });

  factory AskUserRequestData.fromJson(Map<String, dynamic> d) =>
      AskUserRequestData(
        id: d['id'] as String? ?? '',
        title: d['title'] as String? ?? '',
        questions: (d['questions'] as List<dynamic>?)
                ?.map(
                    (q) => AskUserQuestion.fromJson(q as Map<String, dynamic>))
                .toList() ??
            [],
      );
}

class AskUserAnswer {
  final String questionId;
  final List<String> choiceIds;
  final String freeformText;

  AskUserAnswer({
    required this.questionId,
    this.choiceIds = const [],
    this.freeformText = '',
  });

  Map<String, dynamic> toJson() => {
        'question_id': questionId,
        'choice_ids': choiceIds,
        'freeform_text': freeformText,
      };

  factory AskUserAnswer.fromJson(Map<String, dynamic> d) => AskUserAnswer(
        questionId: d['question_id'] as String? ?? '',
        choiceIds: (d['choice_ids'] as List<dynamic>?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        freeformText: d['freeform_text'] as String? ?? '',
      );
}

class AskUserResponseData {
  final String id;
  final String status;
  final List<AskUserAnswer> answers;

  AskUserResponseData({
    required this.id,
    required this.status,
    required this.answers,
  });

  factory AskUserResponseData.fromJson(Map<String, dynamic> d) =>
      AskUserResponseData(
        id: d['id'] as String? ?? '',
        status: d['status'] as String? ?? '',
        answers: (d['answers'] as List<dynamic>?)
                ?.map((a) => AskUserAnswer.fromJson(a as Map<String, dynamic>))
                .toList() ??
            const [],
      );
}

/// ─── Sub-agent / Teammate ───

class SubagentSpawnData {
  final String agentId;
  final String name;
  final String task;
  final String color;
  final String parentId;

  SubagentSpawnData({
    required this.agentId,
    required this.name,
    required this.task,
    this.color = '#4CAF50',
    this.parentId = '',
  });

  factory SubagentSpawnData.fromJson(Map<String, dynamic> d) =>
      SubagentSpawnData(
        agentId: d['agent_id'] as String? ?? '',
        name: d['name'] as String? ?? '',
        task: d['task'] as String? ?? '',
        color: d['color'] as String? ?? '#4CAF50',
        parentId: d['parent_id'] as String? ?? '',
      );
}

class SubagentTextData {
  final String agentId;
  final String id;
  final String chunk;
  final bool done;

  SubagentTextData({
    required this.agentId,
    required this.id,
    required this.chunk,
    required this.done,
  });

  factory SubagentTextData.fromJson(Map<String, dynamic> d) => SubagentTextData(
        agentId: d['agent_id'] as String? ?? '',
        id: d['id'] as String? ?? '',
        chunk: d['chunk'] as String? ?? '',
        done: d['done'] as bool? ?? false,
      );
}

class SubagentReasoningData {
  final String agentId;
  final String id;
  final String chunk;
  final bool done;

  SubagentReasoningData({
    required this.agentId,
    required this.id,
    required this.chunk,
    required this.done,
  });

  factory SubagentReasoningData.fromJson(Map<String, dynamic> d) =>
      SubagentReasoningData(
        agentId: d['agent_id'] as String? ?? '',
        id: d['id'] as String? ?? '',
        chunk: d['chunk'] as String? ?? '',
        done: d['done'] as bool? ?? false,
      );
}

class SubagentStatusData {
  final String agentId;
  final String status;
  final String message;

  SubagentStatusData({
    required this.agentId,
    required this.status,
    required this.message,
  });

  factory SubagentStatusData.fromJson(Map<String, dynamic> d) =>
      SubagentStatusData(
        agentId: d['agent_id'] as String? ?? '',
        status: d['status'] as String? ?? '',
        message: d['message'] as String? ?? '',
      );
}

class SubagentCompleteData {
  final String agentId;
  final String name;
  final String summary;
  final bool success;

  SubagentCompleteData({
    required this.agentId,
    required this.name,
    required this.summary,
    required this.success,
  });

  factory SubagentCompleteData.fromJson(Map<String, dynamic> d) =>
      SubagentCompleteData(
        agentId: d['agent_id'] as String? ?? '',
        name: d['name'] as String? ?? '',
        summary: d['summary'] as String? ?? '',
        success: d['success'] as bool? ?? false,
      );
}

class SubagentToolCallData {
  final String agentId;
  final String toolId;
  final String toolName;
  final String displayName;
  final String args;
  final String detail;

  SubagentToolCallData({
    required this.agentId,
    required this.toolId,
    required this.toolName,
    this.displayName = '',
    this.args = '',
    this.detail = '',
  });

  factory SubagentToolCallData.fromJson(Map<String, dynamic> d) =>
      SubagentToolCallData(
        agentId: d['agent_id'] as String? ?? '',
        toolId: d['tool_id'] as String? ?? '',
        toolName: d['tool_name'] as String? ?? '',
        displayName: d['display_name'] as String? ?? '',
        args: d['args'] as String? ?? '',
        detail: d['detail'] as String? ?? '',
      );
}

class SubagentToolResultData {
  final String agentId;
  final String toolId;
  final String toolName;
  final String result;
  final String summary;
  final String payload;
  final String payloadMode;
  final bool isError;

  SubagentToolResultData({
    required this.agentId,
    required this.toolId,
    required this.toolName,
    required this.result,
    this.summary = '',
    this.payload = '',
    this.payloadMode = '',
    this.isError = false,
  });

  factory SubagentToolResultData.fromJson(Map<String, dynamic> d) =>
      SubagentToolResultData(
        agentId: d['agent_id'] as String? ?? '',
        toolId: d['tool_id'] as String? ?? '',
        toolName: d['tool_name'] as String? ?? '',
        result: d['result'] as String? ?? '',
        summary: d['summary'] as String? ?? '',
        payload: d['payload'] as String? ?? '',
        payloadMode: d['payload_mode'] as String? ?? '',
        isError: d['is_error'] as bool? ?? false,
      );
}

class ErrorData {
  final String message;
  final String code;

  ErrorData({required this.message, required this.code});

  factory ErrorData.fromJson(Map<String, dynamic> d) => ErrorData(
        message: d["message"] as String? ?? "",
        code: d["code"] as String? ?? "",
      );
}
